api/service.go

```diff
@@ -514,6 +525,199 @@ func (s *service) RemoveAppToken(ctx context.Context, req *pb.RemoveAppTokenRequ
+			default:
+				sendErr(fmt.Errorf("invalid request"))
+				return
+			}
+		}
+	}()
+
+	var size string
+	eventCh := make(chan interface{})
+	defer close(eventCh)
+	go func() {
+		for e := range eventCh {
+			event, ok := e.(*iface.AddEvent)
+			if !ok {
+				log.Error("unexpected event type")
+				continue
+			}
+			if event.Path == nil { // This is a progress event
+				sendEvent(&pb.AddFileReply_Event{
+					Name:  event.Name,
+					Bytes: event.Bytes,
+				})
+			} else {
+				size = event.Size // Store size for use in the final response
+			}
+		}
+	}()
+
+	pth, err := s.ipfsClient.Unixfs().Add(
+		server.Context(),
+		ipfsfiles.NewReaderFile(reader),
+		options.Unixfs.Pin(true),
+		options.Unixfs.Progress(true),
+		options.Unixfs.Events(eventCh))
+	if err != nil {
+		return err
+	}
+
+	_, err = s.collections.Files.Create(server.Context(), pth, name, proj.ID)
+	if err != nil && !strings.Contains(err.Error(), store.ErrUniqueExists.Error()) {
+		return err
+	}
+
+	sendEvent(&pb.AddFileReply_Event{
+		Path: pth.String(),
+		Size: size,
+	})
+
+	log.Debugf("stored file with path: %s", pth.String())
+	return nil
+}
+
+// GetFile handles a get file request.
+func (s *service) GetFile(ctx context.Context, req *pb.GetFileRequest) (*pb.GetFileReply, error) {
+	log.Debugf("received get file request")
+
+	scope, ok := ctx.Value(reqKey("scope")).(string)
+	if !ok {
+		log.Fatal("scope required")
+	}
+	file, err := s.getFileWithScope(ctx, req.Path, scope)
+	if err != nil {
+		return nil, err
+	}
+
+	return fileToPbFile(file), nil
+}
+
+func fileToPbFile(file *c.File) *pb.GetFileReply {
+	return &pb.GetFileReply{
+		ID:        file.ID,
+		Path:      file.Path,
+		Name:      file.Name,
+		ProjectID: file.ProjectID,
+		Created:   file.Created,
+	}
+}
+
+// ListFiles handles a list files request.
+func (s *service) ListFiles(ctx context.Context, req *pb.ListFilesRequest) (*pb.ListFilesReply, error) {
+	log.Debugf("received list files request")
+
+	scope, ok := ctx.Value(reqKey("scope")).(string)
+	if !ok {
+		log.Fatal("scope required")
+	}
+	proj, err := s.getProjectForScope(ctx, req.ProjectID, scope)
+	if err != nil {
+		return nil, err
+	}
+
+	files, err := s.collections.Files.List(ctx, proj.ID)
+	if err != nil {
+		return nil, err
+	}
+	list := make([]*pb.GetFileReply, len(files))
+	for i, file := range files {
+		list[i] = fileToPbFile(file)
+	}
+
+	return &pb.ListFilesReply{List: list}, nil
+}
+
+// RemoveFile handles a remove file request.
+func (s *service) RemoveFile(ctx context.Context, req *pb.RemoveFileRequest) (*pb.RemoveFileReply, error) {
+	log.Debugf("received remove file request")
+
+	scope, ok := ctx.Value(reqKey("scope")).(string)
+	if !ok {
+		log.Fatal("scope required")
+	}
+	file, err := s.getFileWithScope(ctx, req.Path, scope)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = s.collections.Files.Delete(ctx, file.ID); err != nil {
+		return nil, err
+	}
+
+	if err = s.ipfsClient.Pin().Rm(ctx, path.New(req.Path)); err != nil {
```

Comment:
Unpin.

Reply:
Thought: I guess GC is mainly the responsibility of the hosting IPFS actor.